crates/ruff_linter/src/rules/ruff/rules/control_var_used_after_block.rs

```diff
@@ -0,0 +1,143 @@
+        } = self;
+
+        format!("{block_kind} variable {control_var_name} is used outside of block")
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+enum BlockKind {
+    For,
+}
+
+impl fmt::Display for BlockKind {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            BlockKind::For => fmt.write_str("`for` loop"),
+        }
+    }
+}
+
+/// Based on wemake-python-styleguide (WPS441) to forbid control variables after the block body.
+pub(crate) fn control_var_used_after_block(
+    checker: &Checker,
+    scope: &Scope,
+    diagnostics: &mut Vec<Diagnostic>,
+) {
+    // Keep track of node_ids we know are used in the same block. This helps us when
+    // people use the same variable name in multiple blocks.
+    let mut known_good_reference_node_ids: Vec<NodeId> = Vec::new();
+
+    let all_bindings: Vec<(&str, BindingId)> = scope.all_bindings().collect();
+    // We want to reverse the bindings so that we iterate in source order and shadowed
+    // bindings come first.
+    let reversed_bindings = all_bindings.iter().rev();
+
+    // Find for-loop variable bindings
+    for (&name, binding) in reversed_bindings
+        .map(|(name, binding_id)| (name, checker.semantic().binding(*binding_id)))
+        .filter_map(|(name, binding)| {
+            if binding.kind.is_loop_var() {
+                return Some((name, binding));
+            }
+
+            None
+        })
+    {
+        let binding_statement = binding.statement(checker.semantic()).unwrap();
+        let binding_source_node_id = binding.source.unwrap();
+        // The node_id of the for-loop that contains the binding
+        let binding_statement_id = checker.semantic().statement_id(binding_source_node_id);
+
+        // Loop over the references of those bindings to see if they're in the same block-scope
+        for reference in binding.references() {
+            let reference = checker.semantic().reference(reference);
+            let reference_node_id = reference.expression_id().unwrap();
+
+            // Skip any reference that come before the control var binding in the source
+            // order, skip it because people can assign and use the same variable name
+            // above the block.
+            if reference.range().end() < binding_statement.range().start() {
+                continue;
+            }
+
+            // Traverse the hierarchy and look for a block match
+            let statement_hierarchy: Vec<NodeId> = checker
+                .semantic()
+                .parent_statement_ids(reference_node_id)
+                .collect();
+
+            let mut is_used_in_block = false;
+            for ancestor_node_id in statement_hierarchy {
+                if binding_statement_id == ancestor_node_id {
+                    is_used_in_block = true;
+                    known_good_reference_node_ids.push(reference_node_id);
+                    break;
+                }
+            }
```

Comment:
Can we add an example where the for loop contains a nested function and it access the variable

```python
for x in range(0..10):
	def test(): 
		print(x)

	class Test:
		a = x
```

What's the expected output in this case?

Reply:
Added :white_check_mark: 

No violations are thrown for that example. I think that's as expected given this sort of usage :thinking: 

```python
for x in range(0, 10):
    def test(): 
        print(x)
        
    test()
        
    class Test:
        a = x
        
    foo = Test()
    print(f"foo {foo.a}")
```